src/flint/flint_base/flint_base.pyx

```diff
@@ -114,16 +119,120 @@ cdef class flint_poly(flint_elem):
                 v = - fac[0]
                 roots.append((v, m))
         return roots
-    
+
     def complex_roots(self):
         raise AttributeError("Complex roots are not supported for this polynomial")
 
 
+cdef class flint_mpoly_context(flint_elem):
+    """
+    Base class for multivariate ring contexts
+    """
+
+    _ctx_cache = None
+
+    def __cinit__(self):
+        self._init = False
+
+    def __init__(self, int nvars, names):
+        if nvars < 0:
+            raise ValueError("cannot have a negative amount of variables")
+        elif len(names) != nvars:
+            raise ValueError("number of variables must match lens of variable names")
+        self.py_names = tuple(bytes(name, 'utf-8') if not isinstance(name, bytes) else name for name in names)
+        self.c_names = <char**>libc.stdlib.malloc(nvars * sizeof(char *))
+        self._init = True
+        for i in range(nvars):
+            self.c_names[i] = self.py_names[i]
```

Comment:
I'm still suspicious of this assignment. What exactly is the generated C code doing?

Reply:
Here's that snippet I posted earlier, de-obfuscated a bit
```c
  for (i = 0; i < __pyx_t_6; i+=1) {
    /* "flint/flint_base/flint_base.pyx":146
 *         self._init = True
 *         for i in range(nvars):
 *             self.c_names[i] = self.py_names[i]             # <<<<<<<<<<<<<<
 * 
 * 
 */
    py_name = __Pyx_GetItemInt(self->py_names, i, int, 1, __Pyx_PyInt_From_int, 0, 1, 1);
    if (unlikely(!py_name)) __PYX_ERR(0, 146, __pyx_L1_error)
        __Pyx_GOTREF(py_name);
        
    __pyx_t_8 = __Pyx_PyObject_AsString(py_name);
    
    if (unlikely((!__pyx_t_8) && PyErr_Occurred())) __PYX_ERR(0, 146, __pyx_L1_error)
        (self->c_names[i]) = __pyx_t_8;
        
    __Pyx_DECREF(py_name); py_name = 0;
  }
```

It seems like it's grabbing the Python object at a particular index, and calling `__Pyx_PyObject_AsString` on it, handling errors when appropriate.

`__Pyx_PyObject_AsString` is a function defined by Cython,
```c
static CYTHON_INLINE const char* __Pyx_PyObject_AsString(PyObject* o) {
    Py_ssize_t ignore;
    return __Pyx_PyObject_AsStringAndSize(o, &ignore);
}
```
`__Pyx_PyObject_AsStringAndSize` is defined similarly,
```c
static CYTHON_INLINE const char* __Pyx_PyObject_AsStringAndSize(PyObject* o, Py_ssize_t *length) {
#if __PYX_DEFAULT_STRING_ENCODING_IS_ASCII || __PYX_DEFAULT_STRING_ENCODING_IS_DEFAULT
    if (
#if PY_MAJOR_VERSION < 3 && __PYX_DEFAULT_STRING_ENCODING_IS_ASCII
            __Pyx_sys_getdefaultencoding_not_ascii &&
#endif
            PyUnicode_Check(o)) {
        return __Pyx_PyUnicode_AsStringAndSize(o, length);
    } else
#endif
#if (!CYTHON_COMPILING_IN_PYPY && !CYTHON_COMPILING_IN_LIMITED_API) || (defined(PyByteArray_AS_STRING) && defined(PyByteArray_GET_SIZE))
    if (PyByteArray_Check(o)) {
        *length = PyByteArray_GET_SIZE(o);
        return PyByteArray_AS_STRING(o);
    } else
#endif
    {
        char* result;
        int r = PyBytes_AsStringAndSize(o, &result, length);
        if (unlikely(r < 0)) {
            return NULL;
        } else {
            return result;
        }
    }
}
```
We're definitely not compiling for Python 2 so we can ignore that bit, I'm not sure whether `(defined(PyByteArray_AS_STRING) && defined(PyByteArray_GET_SIZE)` will be true but I believe it will be. 

Those functions are part of the CPython API, I've copied their docs here. `PyByteArray_GET_SIZE` and `PyByteArray_AS_STRING` are macros for their function counter parts that don't check pointers
> int PyByteArray_Check(PyObject *o)
>
>    Return true if the object o is a bytearray object or an instance of a subtype of the bytearray type. This function always succeeds.
> https://docs.python.org/3/c-api/bytearray.html#c.PyByteArray_Check

> Py_ssize_t PyByteArray_Size(PyObject *bytearray)
>     Part of the Stable ABI.
> 
>     Return the size of bytearray after checking for a NULL pointer.
> https://docs.python.org/3/c-api/bytearray.html#c.PyByteArray_Size

> char *PyByteArray_AsString(PyObject *bytearray)
>    Part of the Stable ABI.
>
>    Return the contents of bytearray as a char array after checking for a NULL pointer. The returned array always has an extra null byte appended.
> https://docs.python.org/3/c-api/bytes.html#c.PyBytes_AsStringAndSize


> int PyBytes_AsStringAndSize(PyObject *obj, char **buffer, Py_ssize_t *length)
>    Part of the Stable ABI.
>
>    Return the null-terminated contents of the object obj through the output variables buffer and length. Returns 0 on success.
>
>    If length is NULL, the bytes object may not contain embedded null bytes; if it does, the function returns -1 and a ValueError is raised.
>
>    The buffer refers to an internal buffer of obj, which includes an additional null byte at the end (not counted in length). The data must not be modified in any way, unless the object was just created using PyBytes_FromStringAndSize(NULL, size). It must not be deallocated. If obj is not a bytes object at all, PyBytes_AsStringAndSize() returns -1 and raises TypeError.
> https://docs.python.org/3/c-api/bytes.html#c.PyBytes_AsStringAndSize

Unfortunately the byte version doesn't have the same guarantees as the string version, but looking at the source code from `bytearrayobject.h`
```c
static inline char* PyByteArray_AS_STRING(PyObject *op)
{
    PyByteArrayObject *self = _PyByteArray_CAST(op);
    if (Py_SIZE(self)) {
        return self->ob_start;
    }
    return _PyByteArray_empty_string;
}
#define PyByteArray_AS_STRING(self) PyByteArray_AS_STRING(_PyObject_CAST(self))
```
It looks like it does not perform a new allocation.

So, that Cython snippet should be grabbing a pointer to the internal buffers of the bytes objects and storing in it the array `c_names`. We are (and should only be) freeing `c_names` itself, but not the underlying data.

Reply:
Okay, that makes sense. I'm unsure what is causing the Windows crash though. I'll try to take a closer look when I can.